src/sacn/dmx_merger.c

```diff
@@ -258,10 +576,89 @@ const SacnDmxMergerSource* sacn_dmx_merger_get_source(sacn_dmx_merger_t merger,
  * \return #kEtcPalErrSys: An internal library or system call error occurred.
  */
 etcpal_error_t sacn_dmx_merger_update_source_data(sacn_dmx_merger_t merger, sacn_source_id_t source, uint8_t priority,
-                                                  const uint8_t* new_values, size_t new_values_count, 
+                                                  const uint8_t* new_values, size_t new_values_count,
                                                   const uint8_t* address_priorities, size_t address_priorities_count)
 {
-  return kEtcPalErrNotImpl;
+  // Verify module initialized.
+  if (!sacn_initialized())
+  {
+    return kEtcPalErrNotInit;
+  }
+
+  // Validate merger.
+  if (merger == SACN_DMX_MERGER_INVALID)
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  // Validate source.
+  if (source == SACN_DMX_MERGER_SOURCE_INVALID)
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  // Validate new_values and new_values_count.
+  if (((new_values != NULL) && (new_values_count == 0)) || ((new_values == NULL) && (new_values_count != 0)))
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  if (new_values_count > DMX_ADDRESS_COUNT)
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  // Validate priority.
+  if (!UNIVERSE_PRIORITY_VALID(priority))
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  // Validate address_priorities and address_priorities_count.
+  if (((address_priorities != NULL) && (address_priorities_count == 0)) ||
+      ((address_priorities == NULL) && (address_priorities_count != 0)))
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  if (address_priorities_count > DMX_ADDRESS_COUNT)
+  {
+    return kEtcPalErrInvalid;
+  }
+
+  // Look up the merger state.
+  MergerState* merger_state = etcpal_rbtree_find(&mergers, &merger);
+
+  if (merger_state == NULL)
+  {
+    return kEtcPalErrNotFound;
+  }
+
+  // Look up the source state.
+  SourceState* source_state = etcpal_rbtree_find(&merger_state->source_state_lookup, &source);
+
+  if (source_state == NULL)
+  {
+    return kEtcPalErrNotFound;
+  }
```

Comment:
I just made changes that apply these rules throughout the code. I did make some exceptions with combining conditions, where it seemed to be more readable to have separate if statements instead of one big if condition block. But let me know if you'd still prefer them to be merged into one if.